### ISSUE TITLE
Possible future examples

### ISSUE THREAD

**Issue by lschmelzeisen** [knowledge-graphs-tutorial/examples#7]
The goal of this issue is to serve as an overview of possible future examples or calculations that we could include in this repository or in the appendix of the extended paper.

I created this list while reading through the current draft for the ACM CSUR submission, and it is structured after those sections. For sections that do not appear in this list, I feel that they are already fully covered, or that adding something would likely be out of scope for the paper. But, feel free to comment with more suggestions!

**4.1.1 Graph Algorithms**

- [ ] Scripts / formulas for...
  - [ ] ... calculating *node* and *edge centrality* with different metrics (degree, betweenness, closenss, Eigenvector, PageRank, HITS, Katz, etc.)
  - [ ] ... *detecting communities* with different algorithms (minimum-cut algorithm, label propagation, Louvain modularity, etc.)
  - [ ] ... perform *connectivity analysis* with multiple techniques (graph density, k-connectivity, strongly connected components, weakly connected componentsm, spanning tree, minimum cuts, etc.)
  - [ ] ... calculate *node similarity* with different metrics (structural equivalence, random walks, diffusion kernels, etc.)
  - [ ] ... calculate *graph summaries* (e.g., via quotient graphs)

**4.1.2 Graph Processing Frameworks**
- [ ] Step-by-step calculation of PageRank values according to figure 17
- [ ] Implementation of the PageRank example in at least one of the mentioned graph parallel frameworks (Apache Spark (GraphX), GraphLab, Pregel, Signal-Collect, Shark)

**4.2 Knowledge Graph Embeddings**
- [ ] Full *one hot embedding* for one of the example graphs (figure 1, figure 15, or figure 18 (a))

**4.2.1 Translational models**
- [ ] Code for training a TransE embedding for one of the examples graphs using some framework (e.g., PyKeen or OpenKE)
- [ ] Actual trained embedding of figure 18 (a) together with visualization of projection that looks somewhat similar to figure 18 (b) and (c)

**4.2.2 Tensor decomposition models**
- [ ] Toy calculation that decomposes a given matrix with Chilean city temperatures per month into 2 (or more) vectors

**Comment by aidhog** [comment]
Would indeed be nice to have these examples.

Regarding 4.1.2, note that we link to such examples on the main page https://github.com/knowledge-graphs-tutorial/examples for Giraph, GraphX and Signal/Collect. Not sure it makes sense to import those same examples into this project.

**Comment by aidhog** [comment]
Just to add, what would be _really_ fantastic would be Jupyter notebooks for Section 4.

**Comment by lschmelzeisen** [comment]
> Regarding 4.1.2, note that we link to such examples on the main page https://github.com/knowledge-graphs-tutorial/examples for Giraph, GraphX and Signal/Collect. Not sure it makes sense to import those same examples into this project.

Yes, I think we have enough notes to make it possible to figure out how to do PageRank. Still, I feel that for beginners there would be value in providing short implementations for one of the examples graphs over just linking to API references. Of course, something like this comes with the added bloat of specifying dependencies, how to build, etc. Else, I probably would have already done it.
